File: skill_registry/runtimes/skill_runner.js
/**
 * 通用 Skill 执行器（Node.js）— 在 skill 的独立 node_modules 中被调用。
 *
 * 协议（JSON over stdio，与 Python runner 一致）：
 *   输入（stdin 一行 JSON）：
 *     {"module_path": "test_node_skill.js", "function_name": "echo", "arguments": {"text": "hello"}}
 *   输出（stdout 一行 JSON）：
 *     {"success": true, "result": "NodeJS echo: hello"}
 *
 * module_path 可以是：
 *   - 绝对路径
 *   - 相对于 skill_registry 目录的路径
 *   - 相对于 cwd（node env 目录）的路径
 */

import * as path from "path";
import * as fs from "fs";
import * as url from "url";
import * as readline from "readline";

const __filename_runner = url.fileURLToPath(import.meta.url);
const __dirname_runner = path.dirname(__filename_runner);
const _SKILL_REGISTRY_DIR = path.dirname(__dirname_runner);
const _AGENT_DIR = path.dirname(_SKILL_REGISTRY_DIR);

function resolveModulePath(modulePath) {
    // 1. 绝对路径
    if (path.isAbsolute(modulePath) && fs.existsSync(modulePath)) {
        return path.toNamespacedPath(modulePath);
    }
    // 2. 相对于 skill_registry 目录
    const skillRegPath = path.join(_SKILL_REGISTRY_DIR, modulePath);
    if (fs.existsSync(skillRegPath)) {
        return path.toNamespacedPath(skillRegPath);
    }
    // 3. 相对于 agent 根目录
    const agentPath = path.join(_AGENT_DIR, modulePath);
    if (fs.existsSync(agentPath)) {
        return path.toNamespacedPath(agentPath);
    }
    // 4. 相对于 cwd
    const cwdPath = path.resolve(modulePath);
    if (fs.existsSync(cwdPath)) {
        return path.toNamespacedPath(cwdPath);
    }
    // 5. 原样返回（让 Node 尝试解析）
    return modulePath;
}

async function handleRequest(req) {
    const modulePath = req.module_path || "";
    const functionName = req.function_name || "";
    const args = req.arguments || {};

    if (!modulePath || !functionName) {
        return { success: false, error: "module_path 和 function_name 不能为空" };
    }

    try {
        const resolvedPath = resolveModulePath(modulePath);
        const fullPath = path.toNamespacedPath(resolvedPath);
        const moduleUrl = url.pathToFileURL(fullPath).href;

        const module = await import(moduleUrl);
        const func = module[functionName] || module.default?.[functionName];
        if (typeof func !== "function") {
            return { success: false, error: `函数 '${functionName}' 不存在于模块 '${modulePath}'` };
        }

        const result = await Promise.resolve(func(args));
        return { success: true, result };
    } catch (e) {
        return { success: false, error: `${e.constructor?.name || "Error"}: ${e.message}` };
    }
}

async function main() {
    const rl = readline.createInterface({ input: process.stdin });

    const line = await new Promise((resolve) => {
        rl.once("line", resolve);
    });

    if (!line || !line.trim()) {
        process.stdout.write(JSON.stringify({ success: false, error: "空输入" }) + "\n");
        return;
    }

    let req;
    try {
        req = JSON.parse(line.trim());
    } catch (e) {
        process.stdout.write(JSON.stringify({ success: false, error: `JSON 解析失败: ${e.message}` }) + "\n");
        return;
    }

    const resp = await handleRequest(req);
    process.stdout.write(JSON.stringify(resp) + "\n");
}

main();

File: skill_registry/runtimes/skill_runner.py
"""通用 Skill 执行器 — 在 skill 的独立 venv 中被调用。

协议（JSON over stdio）：
  输入（stdin 一行 JSON）：
    {"module_path": "mymodule", "function_name": "execute", "arguments": {"text": "hello"}}
  输出（stdout 一行 JSON）：
    {"success": true, "result": "Hello, world!"}
  或
    {"success": false, "error": "ModuleNotFoundError: No module named 'xxx'"}

用法：
  venv_python skill_runner.py < stdin_request.json
"""
import json
import sys
import os
import asyncio

# 把项目根目录和 skill_registry 目录加入 sys.path，使 venv python 能找到 skill 模块
_RUNNER_DIR = os.path.dirname(os.path.abspath(__file__))        # skill_registry/runtimes/
_SKILL_REGISTRY_DIR = os.path.dirname(_RUNNER_DIR)               # skill_registry/
_AGENT_DIR = os.path.dirname(_SKILL_REGISTRY_DIR)               # 项目根目录
for _p in (_AGENT_DIR, _SKILL_REGISTRY_DIR):
    if _p not in sys.path:
        sys.path.insert(0, _p)


def handle_request(req: dict) -> dict:
    module_path = req.get("module_path", "")
    function_name = req.get("function_name", "")
    arguments = req.get("arguments", {})

    if not module_path or not function_name:
        return {"success": False, "error": "module_path 和 function_name 不能为空"}

    try:
        import importlib
        module = importlib.import_module(module_path)
        func = getattr(module, function_name, None)
        if func is None:
            return {"success": False, "error": f"函数 '{function_name}' 不存在于模块 '{module_path}'"}

        if asyncio.iscoroutinefunction(func):
            result = asyncio.run(func(**arguments))
        else:
            result = func(**arguments)

        return {"success": True, "result": result}

    except Exception as e:
        return {"success": False, "error": f"{type(e).__name__}: {e}"}


def main():
    try:
        sys.stdout.reconfigure(encoding="utf-8", errors="replace")
    except Exception:
        pass

    line = sys.stdin.readline()
    if not line or not line.strip():
        print(json.dumps({"success": False, "error": "空输入"}))
        sys.stdout.flush()
        return

    try:
        req = json.loads(line.strip())
    except json.JSONDecodeError as e:
        print(json.dumps({"success": False, "error": f"JSON 解析失败: {e}"}))
        sys.stdout.flush()
        return

    resp = handle_request(req)
    print(json.dumps(resp, ensure_ascii=False, default=str))
    sys.stdout.flush()


if __name__ == "__main__":
    main()
